model/map/src/main/java/org/keycloak/models/map/client/MapClientProvider.java

```diff
@@ -139,15 +139,19 @@ public void unregisterNode(String nodeHost) {
     public ClientModel addClient(RealmModel realm, String id, String clientId) {
         LOG.tracef("addClient(%s, %s, %s)%s", realm, id, clientId, getShortStackTrace());
 
+        if (id != null && tx.read(id) != null) {
+            throw new ModelDuplicateException("Client with same id exists: " + id);
+        }
+        if (clientId != null && getClientByClientId(realm, clientId) != null) {
+            throw new ModelDuplicateException("Client with same clientId in realm " + realm.getName() + " exists: " + clientId);
+        }
+
         MapClientEntity entity = new MapClientEntityImpl();
         entity.setId(id);
         entity.setRealmId(realm.getId());
         entity.setClientId(clientId);
         entity.setEnabled(true);
         entity.setStandardFlowEnabled(true);
-        if (id != null && tx.read(id) != null) {
-            throw new ModelDuplicateException("Client exists: " + id);
-        }
         entity = tx.create(entity);
         if (clientId == null) {
```

Comment:
Hi @vramik - I had a look at that piece of code as well and found the following: 

When id is null, it will be set in tx.create(entity) and the entity returned will have the id set. 
Then, when clientId is null, the newly created id is used to set the clientId in the entity. See JpaClientMapKeycloakTransaction below for the details. I tried to do this in the UI, and actually wasn't able to create a client without a clientId (it is a mandatory field).

Therefore, clientId can't remain null. 

One could push down the code to set the clientId down to the JpaClientMapKeycloakTransaction as well so that it is closer to setting and creating the id. But then it would move to the storage specific code, and then all providers would need to do it. Therefore I'd like to keep it as (weird as it) is. 

``` 
    @Override
    public MapClientEntity create(MapClientEntity mapEntity) {
        JpaClientEntity jpaEntity = (JpaClientEntity) CLONER.from(mapEntity);
        if (mapEntity.getId() == null) {
            jpaEntity.setId(UUIDKey.INSTANCE.yieldNewUniqueKey().toString());
        }
        jpaEntity.setEntityVersion(SUPPORTED_VERSION_CLIENT);
        em.persist(jpaEntity);
        return jpaEntity;
    }
```

Reply:
You're right. I see it now, I've missed the fact that `tx.create(entity);` is called before checking the `clientId`. Please consider this this as resolved.